### PR TITLE
[12.0][ADD] stock_picking_tier_validation

### PR DESCRIPTION
**Pull request by TimLai125**
This module extends the functionality of Pickings to support a tier validation process.
The original author the module is https://github.com/sistemasdedatos/odoo-addons, yet the module is generic similar to [purchase_request_tier_validation](https://github.com/OCA/purchase-workflow/tree/12.0/purchase_request_tier_validation).

### CODE REVIEW

File: stock_picking_tier_validation/models/stock_picking.py
Comment (yostashiro) [inline stock_picking_tier_validation/models/stock_picking.py]:
```suggestion
    _state_from = ["draft", "waiting", "confirmed", "assigned"]
    _state_to = ["done"]
```
@TimLai125 Shouldn't this part be like this?

Comment (TimLai125) [reply stock_picking_tier_validation/models/stock_picking.py]: Yes, we should include other states as well since picking created from orders (PO/SO etc.) may already be proceed to assigned state in the beginning. Thanks for the comment.

**Comment by TimLai125** [comment]
base_tier_validation does not support state field as a compute field. Closing the PR now, please feel free to re-open it once there is a workaround.

**Comment by TimLai125** [comment]
> base_tier_validation does not support state field as a compute field. Closing the PR now, please feel free to re-open it once there is a workaround.

I am adding an extra constrains method to the stock move model as workaround. Picking that tries to move to the `_state_to` without reviewers will be blocked by the constrains.

File: stock_picking_tier_validation/views/stock_picking_views.xml
Comment (yostashiro) [inline stock_picking_tier_validation/views/stock_picking_views.xml:98]:
```suggestion
                    help="Pickings to review"
```

File: stock_picking_tier_validation/views/stock_picking_views.xml
Comment (yostashiro) [inline stock_picking_tier_validation/views/stock_picking_views.xml:96]:
```suggestion
                    string="Need My Review"
```

File: stock_picking_tier_validation/models/stock_picking.py
Comment (yostashiro) [inline stock_picking_tier_validation/models/stock_picking.py:10]:
```suggestion
    _state_from = ["waiting", "confirmed", "assigned"]
```
Validate button doesn't show when the state is draft.

**Review by yostashiro — CHANGES_REQUESTED** [review]
@TimLai125 I have created this PR https://github.com/OCA/server-ux/pull/230 and once it's merged, the adjustments can be made simpler like below, instead of having the constraint on stock.move.

```.py
from odoo import models


class Picking(models.Model):
    _name = "stock.picking"
    _inherit = ["stock.picking", "tier.validation"]
    _state_from = ["waiting", "confirmed", "assigned"]

    def _check_state_conditions(self, vals):
        self.ensure_one()
        if self.move_line_ids.filtered(lambda x: x.state == "done"):
            return True
        return False
```

**Review by Ryoko04 — APPROVED** [review]
This module makes it possible to approve delivery order, receipt order and transfer order before validating the order.
I ran this module on runbot and worked fine!

**Comment by rousseldenis** [comment]
@TimLai125 Please check runbot warnings :

image

**Pull request closed without merging**
